src/Parser.zig

```diff
@@ -850,12 +850,36 @@ fn typeof(p: *Parser) Error!?Type {
     const l_paren = try p.expectToken(.l_paren);
     if (try p.typeName()) |ty| {
         try p.expectClosing(l_paren, .r_paren);
-        return ty;
+        const typeof_ty = try p.arena.create(Type);
+        typeof_ty.* = .{
+            .data = ty.data,
+            .qual = ty.qual, // todo: exclude register qualifier
+            .specifier = ty.specifier,
+        };
+
+        return Type{
+            .data = .{ .sub_type = typeof_ty },
+            .specifier = .typeof_type,
+        };
     }
     const typeof_expr = try p.parseNoEval(expr);
     try typeof_expr.expect(p);
     try p.expectClosing(l_paren, .r_paren);
-    return typeof_expr.ty;
+
+    const inner = try p.arena.create(Type.VLA);
```

Comment:
It would probably be appropriate to rename `Type.VLA` to `Type.Expr` or something.

Reply:
Renamed it to `Type.Expr` and renamed the members to be (I think) more consistent. It's in its own commit so should be easy to revert or change if you don't like it.